entities and cli removal from ui host  module

tstrych opened the pull request:
THIS PR contains:
- usage of entities is removed from the module and some fixtures ( only some because otherwise, it would be changed across the whole robottelo, I have no time to test such change)
- this is the new proposed change to allow the common use of "default_sat" (now target_sat) objects in fixtures
- refactoring of test host UI module to use API instead cli
- use  fixtures from global modules, instead of local one
- move the local fixtures to global fixtures
- fixes UI puppet related tests (at the bottom of the module)
- new host details are not available by default, so airgun changes were made together with changes in `test_positive_read_details_page_from_new_ui`


I wouldd like to open the discussion about module_user using default_sat in confest.py
https://github.com/SatelliteQE/robottelo/blob/a19eb52cc7ca8a05d55a4240a9354f9fcf7393c9/tests/foreman/ui/conftest.py#L23
and also pytest taxonomy fixtures require the change
https://github.com/SatelliteQE/robottelo/blob/ecbf6f7fe7f348597f8f619ee404106cd8cf0788/pytest_fixtures/component/taxonomy.py
It would break all non-default satellites, `puppet_enabled_sat` is already used in plenty of tests in UI. We need the solution that  @JacobCallahan  proposed and will overwrite the default_sat hostname whenever the non-default sat is needed.

I run the test locally to make sure whether the changed from CLI to API was done correctly. But I didn't run tests where I just change entities to default_sat.  
There will still be a lot of failures, but my intention was to resolve the BoxKeyError and to do some refactoring. 

nailgun PR already merged:
https://github.com/SatelliteQE/nailgun/pull/829

I hope the comparison between test automation results and prt test results will be good enough for this PR to be merged.

tstrych commented on robottelo/host_helpers/cli_factory.py:
```diff
@@ -253,7 +253,8 @@ def __getattr__(self, name):
         """
-        if fields := ENTITY_FIELDS.get(name.replace('make_', '')):
+        fields = ENTITY_FIELDS.get(name.replace('make_', ''))
+        if isinstance(fields, dict):
```
`ENTITY_FIELDS.get` returns `None` if e.g`make_abcdefg` doesn't exist
previous solution when  `ENTITY_FIELDS.get` returns `{}`  which is correct scenario formultiple make_entities didn't met the condition if {} -> false, instead of true

devendra104 commented:
trigger: test-robottelo
pytest: tests/foreman/ui/test_host.py

jyejare commented on tests/foreman/ui/test_host.py:
```diff
@@ -206,27 +197,29 @@ def module_libvirt_domain(module_org, smart_proxy_location, module_proxy, defaul
     # location and dns to it
     _, _, domain = default_sat.hostname.partition('.')
-    domain = entities.Domain().search(query={'search': f'name="{domain}"'})
+    domain = default_sat.api.Domain().search(query={'search': f'name="{domain}"'})
```
Use the `default_domain` fixture from global instead !

jyejare commented on tests/foreman/ui/test_host.py:
```diff
@@ -258,25 +251,25 @@ def module_libvirt_subnet(module_org, smart_proxy_location, module_libvirt_domai
 @pytest.fixture(scope='module')
-def module_lce(module_org):
-    return entities.LifecycleEnvironment(organization=module_org).create()
+def module_lce(module_org, default_sat):
```
Use existing global fixture by modifying that to use default sat.

jyejare commented on tests/foreman/ui/test_host.py:
```diff
@@ -258,25 +251,25 @@ def module_libvirt_subnet(module_org, smart_proxy_location, module_libvirt_domai
 @pytest.fixture(scope='module')
-def module_product(module_org):
-    return entities.Product(organization=module_org).create()
+def module_product(module_org, default_sat):
```
Use existing from global fixtures!

jyejare commented on tests/foreman/ui/test_host.py:
```diff
@@ -258,25 +251,25 @@ def module_libvirt_subnet(module_org, smart_proxy_location, module_libvirt_domai
-def module_repository(os_path, module_product):
-    repo = entities.Repository(product=module_product, url=os_path).create()
-    call_entity_method_with_timeout(entities.Repository(id=repo.id).sync, timeout=3600)
+def module_repository(os_path, module_product, default_sat):
```
Worth as a global fixture !

jyejare commented on tests/foreman/ui/test_host.py:
```diff
@@ -298,9 +291,9 @@ def module_libvirt_media(module_org, smart_proxy_location, os_path, module_os):
+def module_content_view(module_org, module_repository, module_lce, default_sat):
     # Create, Publish and promote CV
-    content_view = entities.ContentView(organization=module_org).create()
+    content_view = default_sat.api.ContentView(organization=module_org).create()
```
a global `module_cv` fixture ?

jyejare commented on tests/foreman/ui/test_host.py:
```diff
@@ -373,16 +362,32 @@ def module_activation_key(manifest_org):
-def remove_vm_on_delete():
-    Settings.set({'name': 'destroy_vm_on_host_delete', 'value': 'true'})
-    assert Settings.list({'search': 'name=destroy_vm_on_host_delete'})[0]['value'] == 'true'
+def remove_vm_on_delete(default_sat):
```
I think you should use `setting_update` fixture here from global settings.py

jyejare commented on tests/foreman/ui/test_host.py:
```diff
@@ -540,11 +545,13 @@ def test_positive_inherit_puppet_env_from_host_group_when_action(session):
-    host = entities.Host(organization=org, location=loc).create()
-    env = entities.Environment(organization=[org], location=[loc]).create()
-    hostgroup = entities.HostGroup(environment=env, organization=[org], location=[loc]).create()
+    org = default_sat.api.Organization().create()
+    loc = default_sat.api.Location().create()
```
Global function_org/function_loc fixtures ?

jyejare reviewed (changes requested):
ACK pending some non blocking improvement comments and a blocking comment ! 

Also, I would recommend you to use as much global fixtures as you can or move the test module level fixtures to global so others can use it !

jyejare commented:
@tstrych What fixture in taxonomy.py, in particular, is problematic! Most fixtures in global fixtures are initially created for default_sat and taxonomy fixtures are one of them.

We are ofcourse waiting for @JacobCallahan PR that replaces on demand sat for fixtures :)

tstrych commented:
Thanks, @jyejare I appreciate the comments. I will work on them during the next week.
I thought we want to replace all occurrences of  `entities` for `default_sat.api`.  (to not cause box key errors), but I think I don't understand properly when it's truly required. Maybe you @jyejare or @JacobCallahan can provide me with more info on this topic.

That's why I was mentioning the taxonomy and contest files. 
To answer your question. All of the fixtures in `taxonomy.py` are still using `entities` instead of the `default_sat`  same for the `confest.py` linked at the description.

JacobCallahan commented on pytest_fixtures/component/smartproxy.py:
```diff
@@ -10,11 +9,13 @@
     smart_proxy = (
-        entities.SmartProxy().search(query={'search': f'name={settings.server.hostname}'})[0].read()
+        default_sat.api.SmartProxy()
+        .search(query={'search': f'name={settings.server.hostname}'})[0]
```
the fixture can use default_sat without an issue. compatibility with destructuve_sat is where we run into issues.

JacobCallahan commented:
@jyejare i think you're referring to #9110 which is something we've not come back around to yet

tstrych commented:
Meanwhile, I fixed some errors mainly puppet UI tests,
and fixed locator in 
https://github.com/SatelliteQE/airgun/pull/696

tstrych commented:
my local tests results, but I also changed things  afterwards, so waiting for PRT

Local test results:
```
collected 42 items / 2 deselected / 40 selected                                                                                                                                              

tests/foreman/ui/test_host.py::test_positive_end_to_end PASSED                                                                                                                         [  2%]
tests/foreman/ui/test_host.py::test_positive_read_from_details_page PASSED                                                                                                             [  5%]
tests/foreman/ui/test_host.py::test_positive_read_from_edit_page PASSED                                                                                                                [  7%]
tests/foreman/ui/test_host.py::test_positive_assign_taxonomies PASSED                                                                                                                  [ 10%]
tests/foreman/ui/test_host.py::test_positive_assign_compliance_policy PASSED                                                                                                           [ 12%]
tests/foreman/ui/test_host.py::test_positive_export FAILED                                                                                                                             [ 15%]
tests/foreman/ui/test_host.py::test_positive_create_with_inherited_params PASSED                                                                                                       [ 17%]
tests/foreman/ui/test_host.py::test_negative_delete_primary_interface PASSED                                                                                                           [ 20%]
tests/foreman/ui/test_host.py::test_positive_view_hosts_with_non_admin_user PASSED                                                                                                     [ 22%]
tests/foreman/ui/test_host.py::test_positive_remove_parameter_non_admin_user PASSED                                                                                                    [ 25%]
tests/foreman/ui/test_host.py::test_negative_remove_parameter_non_admin_user PASSED                                                                                                    [ 27%]
tests/foreman/ui/test_host.py::test_positive_check_permissions_affect_create_procedure PASSED                                                                                          [ 30%]
tests/foreman/ui/test_host.py::test_positive_search_by_parameter PASSED                                                                                                                [ 32%]
tests/foreman/ui/test_host.py::test_positive_search_by_parameter_with_different_values PASSED                                                                                          [ 35%]
tests/foreman/ui/test_host.py::test_positive_search_by_parameter_with_prefix PASSED                                                                                                    [ 37%]
tests/foreman/ui/test_host.py::test_positive_search_by_parameter_with_operator PASSED                                                                                                  [ 40%]
tests/foreman/ui/test_host.py::test_positive_search_with_org_and_loc_context PASSED                                                                                                    [ 42%]
tests/foreman/ui/test_host.py::test_positive_search_by_org PASSED                                                                                                                      [ 45%]
tests/foreman/ui/test_host.py::test_positive_validate_inherited_cv_lce PASSED                                                                                                          [ 47%]
tests/foreman/ui/test_host.py::test_positive_global_registration_form PASSED                                                                                                           [ 50%]
tests/foreman/ui/test_host.py::test_positive_global_registration_end_to_end FAILED                                                                                                     [ 52%]
tests/foreman/ui/test_host.py::test_global_registration_form_populate FAILED                                                                                                           [ 55%]
tests/foreman/ui/test_host.py::test_global_registration_with_capsule_host[rhel7_contenthost0] ERROR                                                                                    [ 57%]
tests/foreman/ui/test_host.py::test_global_registration_with_gpg_repo_and_default_package[rhel7_contenthost0] FAILED                                                                   [ 60%]
tests/foreman/ui/test_host.py::test_global_re_registration_host_with_force_ignore_error_options[rhel7_contenthost0] FAILED                                                             [ 62%]
tests/foreman/ui/test_host.py::test_global_registration_token_restriction[rhel7_contenthost0] FAILED                                                                                   [ 65%]
tests/foreman/ui/test_host.py::test_positive_set_multi_line_and_with_spaces_parameter_value FAILED                                                                                     [ 67%]
tests/foreman/ui/test_host.py::test_positive_bulk_delete_host PASSED                                                                                                                   [ 70%]
tests/foreman/ui/test_host.py::test_positive_gce_provision_end_to_end ERROR                                                                                                            [ 72%]
tests/foreman/ui/test_host.py::test_positive_gce_cloudinit_provision_end_to_end ERROR                                                                                                  [ 75%]
tests/foreman/ui/test_host.py::test_positive_cockpit[rhel7] FAILED                                                                                                                     [ 77%]
tests/foreman/ui/test_host.py::test_positive_cockpit[rhel7_fips] PASSED                                                                                                                [ 80%]
tests/foreman/ui/test_host.py::test_positive_cockpit[rhel8] FAILED                                                                                                                     [ 82%]
tests/foreman/ui/test_host.py::test_positive_cockpit[rhel8_fips] PASSED                                                                                                                [ 85%]
tests/foreman/ui/test_host.py::test_positive_cockpit[rhel9] PASSED                                                                                                                     [ 87%]
tests/foreman/ui/test_host.py::test_positive_cockpit[rhel9_fips] PASSED                                                                                                                [ 90%]
tests/foreman/ui/test_host.py::test_positive_read_details_page_from_new_ui FAILED                                                                                                      [ 92%]
tests/foreman/ui/test_host.py::test_positive_inherit_puppet_env_from_host_group_when_action FAILED                                                                                     [ 95%]
tests/foreman/ui/test_host.py::test_positive_create_with_puppet_class FAILED                                                                                                           [ 97%]
tests/foreman/ui/test_host.py::test_positive_inherit_puppet_env_from_host_group_when_create PASSED                                                                                     [100%]
===================================================== 12 failed, 25 passed, 2 deselected, 448 warnings, 3 errors in 15579.13s (4:19:39) ======================================================
```

I fixed 5 tests from the results above:
```
tests/foreman/ui/test_host.py::test_positive_inherit_puppet_env_from_host_group_when_action
================= 1 passed, 41 deselected in 250.94s (0:04:10) =================

tests/foreman/ui/test_host.py::test_positive_create_with_puppet_class
=========== 1 passed, 41 deselected, 1 warning in 284.70s (0:04:44) ============

tests/foreman/ui/test_host.py::test_positive_export
================= 1 passed, 41 deselected in 110.90s (0:01:50) =================

tests/foreman/ui/test_host.py::test_positive_read_details_page_from_new_ui
================= 1 passed, 41 deselected in 151.12s (0:02:31) =================

tests/foreman/ui/test_host.py::test_positive_set_multi_line_and_with_spaces_parameter_value
================= 1 passed, 41 deselected in 148.60s (0:02:28) =================
```

Remaining, (more time efficient scenarios to fix): 
6 Global registration tests - 
2 Cockpit tests - it seems like a flakiness, need to look closer
2 GCE tests

tstrych commented:
trigger: test-robottelo
pytest: tests/foreman/ui/test_host.py
airgun: 696

tstrych commented:
I would like to get this merged during the next week, before 20 May.

tstrych commented:
depends which one is sooner but default_sat should be replaced by target_sat
 https://github.com/SatelliteQE/robottelo/pull/9562,

tstrych commented:
trigger: test-robottelo
pytest: tests/foreman/ui/test_host.py
airgun: 696

tstrych commented:
trigger: test-robottelo
pytest: tests/foreman/ui/test_host.py
airgun: 696

tstrych commented:
build 799, 28 failed

Awkward spot. 

Because these tests are using a new host UI, 
but meanwhile, the BZ 
https://bugzilla.redhat.com/show_bug.cgi?id=2073421
hit the downstream within the recent snap, which disabled the new host UI by default, so the navigation has to be changed back

The actual airgun PR is still valid because the new host details UI will be enabled in the next version. 

the additional airgun PR, I have already created an issue #695 as I knew this would happen, just didn't know when :)

jyejare commented:
@tstrych Do you want to hold the re-review of this PR? probably you dont want to hold.

tstrych commented:
No, I don't, but I will still create an airgun patch for 6.11 :) to get better results.

tstrych commented:
Run 806: 
23 failed:
```
tests.foreman.ui.test_host.test_positive_assign_taxonomies
tests.foreman.ui.test_host.test_positive_create_with_inherited_params
tests.foreman.ui.test_host.test_global_registration_form_populate
tests.foreman.ui.test_host.test_global_registration_with_capsule_host[rhel7_contenthost0]
tests.foreman.ui.test_host.test_global_registration_with_gpg_repo_and_default_package[rhel7_contenthost0]
tests.foreman.ui.test_host.test_global_registration_upgrade_subscription_manager[rhel7]
tests.foreman.ui.test_host.test_global_registration_upgrade_subscription_manager[rhel7_fips]
tests.foreman.ui.test_host.test_global_registration_upgrade_subscription_manager[rhel8]
tests.foreman.ui.test_host.test_global_registration_upgrade_subscription_manager[rhel8_fips]
tests.foreman.ui.test_host.test_global_registration_upgrade_subscription_manager[rhel9]
tests.foreman.ui.test_host.test_global_registration_upgrade_subscription_manager[rhel9_fips]
tests.foreman.ui.test_host.test_global_re_registration_host_with_force_ignore_error_options[rhel7_contenthost0]
tests.foreman.ui.test_host.test_global_registration_token_restriction[rhel7_contenthost0]
tests.foreman.ui.test_host.test_positive_cockpit[rhel7]
tests.foreman.ui.test_host.test_positive_cockpit[rhel7_fips]
tests.foreman.ui.test_host.test_positive_cockpit[rhel8]
tests.foreman.ui.test_host.test_positive_cockpit[rhel8_fips]
tests.foreman.ui.test_host.test_positive_cockpit[rhel9]
tests.foreman.ui.test_host.test_positive_cockpit[rhel9_fips]
tests.foreman.ui.test_host.test_positive_inherit_puppet_env_from_host_group_when_action
tests.foreman.ui.test_host.test_positive_create_with_puppet_class
tests.foreman.ui.test_host.test_positive_inherit_puppet_env_from_host_group_when_create
tests.foreman.ui.test_host.test_positive_set_multi_line_and_with_spaces_parameter_value
```
2 tests to be fixed ->
tests.foreman.ui.test_host.test_positive_assign_taxonomies
tests.foreman.ui.test_host.test_positive_create_with_inherited_params

other are out of PR scope 
global registration fixes
and other failures are caused by  `default_sat.register_to_dogfood()`, but when tested locally registration has proceeded without a problem

peterdragun commented on pytest_fixtures/component/contentview.py:
```diff
@@ -43,6 +44,18 @@ def module_ak_cv_lce(module_org, module_lce, module_published_cv):
 
 
+@pytest.fixture(scope='module')
+def module_content_view(module_org, module_repository, module_lce, default_sat):
```
as there already is a module_cv fixture would it be better to rename this one to module_cv_repo or something like that to better differentiate the names of those 2 fixtures?

tstrych replied:
I called it `module_cv_repo`, went with logic of `module_ak_cv_lce`

peterdragun commented on pytest_fixtures/component/taxonomy.py:
```diff
@@ -52,6 +52,11 @@ def function_location():
 
+@pytest.fixture(scope='function')
+def function_location_with_org(default_sat):
+    return default_sat.api.Location(organization=[function_org]).create()
```
I think `function_org` should be also in the parameters of this fixture

tstrych replied:
nice catch, I believe this is the only catch related to the test failures made by these changes :+1:

tstrych commented:
Fix for the two failed tests:
```
collected 42 items / 40 deselected / 2 selected
tests/foreman/ui/test_host.py::test_positive_create_with_inherited_params
tests/foreman/ui/test_host.py::test_positive_assign_taxonomies
========== 2 passed, 40 deselected, 13 warnings in 367.04s (0:06:07) ===========
```

tstrych commented:
puppet tests re-run locally with puppet enabled sat inserted this way:
```
from robottelo.hosts import Satellite
session_puppet_enabled_sat = Satellite('fqdn')
```
test run with last commit included [puppet enabled tests using module based proxy location update](https://github.com/SatelliteQE/robottelo/pull/9553/commits/dc75b693ecfe9363aa4d4e3862bc04f571fe1779)

```
tests/foreman/ui/test_host.py::test_positive_inherit_puppet_env_from_host_group_when_action 
tests/foreman/ui/test_host.py::test_positive_create_with_puppet_class
tests/foreman/ui/test_host.py::test_positive_inherit_puppet_env_from_host_group_when_create 
tests/foreman/ui/test_host.py::test_positive_set_multi_line_and_with_spaces_parameter_value
========== 4 passed, 38 deselected, 39 warnings in 560.04s (0:09:20) ===========
```

JacobCallahan commented on pytest_fixtures/component/architecture.py:
```diff
@@ -1,14 +1,15 @@
 
 @pytest.fixture(scope='session')
-def default_architecture():
+def default_architecture(default_sat):
```
all uses of the default_Sat fixture must be replaced with `<scope>_target_sat` after the merge of #9562

peterdragun commented:
As Tomas is on the leave and this PR will fix multiple tests I have rebased this PR to master and replaced all default_sat occurrences with the correct target_sat.

peterdragun commented:
trigger: test-robottelo
pytest: tests/foreman/ui/test_host.py -k 'not (cockpit or registration or gce)'

mshriver reviewed (approved):
I did not review test content in detail, structure and entity class access changes look great.

JacobCallahan commented on pytest_fixtures/component/host.py:
```diff
@@ -8,6 +8,11 @@
 from robottelo.constants import REPOSET
 
 
+@pytest.fixture(scope='function')
```
non-blocking
```suggestion
@pytest.fixture
```

JacobCallahan commented on pytest_fixtures/component/os.py:
```diff
@@ -2,9 +2,11 @@
 import pytest
 from nailgun import entities
 
+from robottelo.config import settings
```
you can move this import into the fixture that needs it

peterdragun replied:
I can but when there will be another fixture that needs settings this will require another change. Is there any reason to move this import to function if it is not conflicting with anything else? I would rather keep imports in one place

peterdragun replied:
Thanks for the explanation, moving into the fixture

vsedmik reviewed (approved):
Focused on the puppet-related changes and all looks good to me. :+1:

peterdragun commented on pytest_fixtures/component/taxonomy.py:
```diff
@@ -53,6 +53,11 @@ def function_location(target_sat):
     return target_sat.api.Location().create()
 
 
+@pytest.fixture(scope='function')
```
:+1: also removing other implicit scopes from this file to keep it consistent
